File: web/src/api/customer.ts
import client, { unwrap, type PageData } from './client'

export interface CustomerItem {
  id: number
  tenantId: number
  displayName: string
  primaryPhone: string
  status: number
  source: string
  remark?: string
  createdAt: string
  updatedAt: string
}

export interface AddressItem {
  id: number
  customerId: number
  contactName: string
  phone: string
  province: string
  city: string
  district: string
  detail: string
  label: string
  isDefault: number
  createdAt: string
  updatedAt: string
}

export interface BindingItem {
  id: number
  customerId: number
  channelType: string
  channelUserId: string
  verified: number
  boundAt?: string
  meta?: string
  createdAt: string
  updatedAt: string
}

export interface CustomerDetail extends CustomerItem {
  addresses: AddressItem[]
  bindings: BindingItem[]
}

export interface DashboardStats {
  customerCount: number
}

export async function fetchDashboardStats() {
  return unwrap<DashboardStats>(await client.get('/dashboard/stats'))
}

export async function listCustomers(params: {
  page?: number
  pageSize?: number
  keyword?: string
  phone?: string
  status?: number
}) {
  return unwrap<PageData<CustomerItem>>(await client.get('/customers', { params }))
}

export async function createCustomer(body: {
  displayName?: string
  primaryPhone: string
  source?: string
  remark?: string
  status?: number
}) {
  return unwrap<CustomerItem>(await client.post('/customers', body))
}

export async function getCustomer(id: number) {
  return unwrap<CustomerDetail>(await client.get(`/customers/${id}`))
}

export async function updateCustomer(id: number, body: Record<string, unknown>) {
  return unwrap<CustomerItem>(await client.put(`/customers/${id}`, body))
}

export async function disableCustomer(id: number) {
  return unwrap<{ disabled: boolean }>(await client.delete(`/customers/${id}`))
}

export async function createAddress(customerId: number, body: Record<string, unknown>) {
  return unwrap<AddressItem>(await client.post(`/customers/${customerId}/addresses`, body))
}

export async function updateAddress(customerId: number, addrId: number, body: Record<string, unknown>) {
  return unwrap<AddressItem>(await client.put(`/customers/${customerId}/addresses/${addrId}`, body))
}

export async function deleteAddress(customerId: number, addrId: number) {
  return unwrap<{ deleted: boolean }>(await client.delete(`/customers/${customerId}/addresses/${addrId}`))
}

export async function createBinding(customerId: number, body: Record<string, unknown>) {
  return unwrap<BindingItem>(await client.post(`/customers/${customerId}/bindings`, body))
}

export async function updateBinding(customerId: number, bindId: number, body: Record<string, unknown>) {
  return unwrap<BindingItem>(await client.put(`/customers/${customerId}/bindings/${bindId}`, body))
}

export async function deleteBinding(customerId: number, bindId: number) {
  return unwrap<{ deleted: boolean }>(await client.delete(`/customers/${customerId}/bindings/${bindId}`))
}

File: internal/dto/customer.go
package dto

type CustomerCreateInput struct {
	DisplayName  string `json:"displayName"`
	PrimaryPhone string `json:"primaryPhone" binding:"required"`
	Source       string `json:"source"`
	Remark       string `json:"remark"`
	Status       *int8  `json:"status"`
}

type CustomerUpdateInput struct {
	DisplayName  *string `json:"displayName"`
	PrimaryPhone *string `json:"primaryPhone"`
	Source       *string `json:"source"`
	Remark       *string `json:"remark"`
	Status       *int8   `json:"status"`
}

type AddressInput struct {
	ContactName string `json:"contactName"`
	Phone       string `json:"phone"`
	Province    string `json:"province"`
	City        string `json:"city"`
	District    string `json:"district"`
	Detail      string `json:"detail"`
	Label       string `json:"label"`
	IsDefault   *int8  `json:"isDefault"`
}

type BindingInput struct {
	ChannelType   string `json:"channelType" binding:"required"`
	ChannelUserID string `json:"channelUserId" binding:"required"`
	Verified      *int8  `json:"verified"`
	Meta          string `json:"meta"`
}

type UpsertByPhoneInput struct {
	TenantID     uint64        `json:"tenantId"`
	Phone        string        `json:"phone" binding:"required"`
	DisplayName  string        `json:"displayName"`
	Source       string        `json:"source"`
	Remark       string        `json:"remark"`
	ChannelType  string        `json:"channelType"`
	ChannelUserID string       `json:"channelUserId"`
	Address      *AddressInput `json:"address"`
}

type UpsertByPhoneResult struct {
	CustomerID uint64       `json:"customerId"`
	Created    bool         `json:"created"`
	Customer   CustomerItem `json:"customer"`
}

type CustomerItem struct {
	ID           uint64 `json:"id"`
	TenantID     uint64 `json:"tenantId"`
	DisplayName  string `json:"displayName"`
	PrimaryPhone string `json:"primaryPhone"`
	Status       int8   `json:"status"`
	Source       string `json:"source"`
	Remark       string `json:"remark,omitempty"`
	CreatedAt    string `json:"createdAt"`
	UpdatedAt    string `json:"updatedAt"`
}

type AddressItem struct {
	ID          uint64 `json:"id"`
	CustomerID  uint64 `json:"customerId"`
	ContactName string `json:"contactName"`
	Phone       string `json:"phone"`
	Province    string `json:"province"`
	City        string `json:"city"`
	District    string `json:"district"`
	Detail      string `json:"detail"`
	Label       string `json:"label"`
	IsDefault   int8   `json:"isDefault"`
	CreatedAt   string `json:"createdAt"`
	UpdatedAt   string `json:"updatedAt"`
}

type BindingItem struct {
	ID            uint64 `json:"id"`
	CustomerID    uint64 `json:"customerId"`
	ChannelType   string `json:"channelType"`
	ChannelUserID string `json:"channelUserId"`
	Verified      int8   `json:"verified"`
	BoundAt       string `json:"boundAt,omitempty"`
	Meta          string `json:"meta,omitempty"`
	CreatedAt     string `json:"createdAt"`
	UpdatedAt     string `json:"updatedAt"`
}

type CustomerDetail struct {
	CustomerItem
	Addresses []AddressItem `json:"addresses"`
	Bindings  []BindingItem `json:"bindings"`
}
